jax/_src/lax/lax.py

```diff
@@ -2811,26 +2812,56 @@ def _broadcast_in_dim_transpose_rule(ct, operand, *dyn_shape,
 
 def _broadcast_in_dim_batch_rule(batched_args, batch_dims, shape,
                                  broadcast_dimensions):
+  # `dyn_shape` is the dynamic portion of the target shape.  `shape`
+  # is the target shape, with `None` for dynamic sections.
+  # broadcast_dimensions gives indices where dimensions of the input
+  # have to go: dimension i of the input becomes dimension
+  # broadcast_dimensions[i] of the output.
   operand, *dyn_shape = batched_args
   operand_bdim, *dyn_shape_bdims = batch_dims
-  if len(dyn_shape) > 1: raise NotImplementedError
-  if (operand_bdim is not None and
-      (not dyn_shape_bdims or dyn_shape_bdims[0] is None)):
-    new_operand = batching.moveaxis(operand, operand_bdim, 0)
-    new_shape = (operand.shape[operand_bdim],) + _merge_dyn_shape(shape, dyn_shape)
+
+  stacked_size = None
+  if operand_bdim is not None:
+    if isinstance(operand_bdim, RaggedAxis):
+      stacked_axis = operand_bdim.stacked_axis
+    else:
+      stacked_axis = operand_bdim
+    new_operand = batching.moveaxis(operand, stacked_axis, 0)
+    if isinstance(operand_bdim, RaggedAxis):
+      stacked_size = operand_bdim.size
+    else:
+      stacked_size = operand.shape[stacked_axis]
     new_broadcast_dimensions = (0,) + tuple(np.add(1, broadcast_dimensions))
-    return broadcast_in_dim(new_operand, new_shape, new_broadcast_dimensions), 0
-  elif (operand_bdim is None and dyn_shape_bdims and
-        dyn_shape_bdims[0] is not None):
-    (d,), (d_bdim,) = dyn_shape, dyn_shape_bdims  # NotImplementedError above
-    assert d_bdim == 0  # must be scalar in the program to be batched
-    bound = d.dtype.bound
-    new_shape = (len(d),) + _merge_dyn_shape(shape, (bound,))
-    out = broadcast_in_dim(operand, new_shape, broadcast_dimensions)
-    idx, = (i for i, s in enumerate(shape) if s is None)
-    return out, batching.RaggedAxis(0, idx+1, d)
   else:
-    raise NotImplementedError  # TODO(mattjj,axch)
+    new_operand = operand
+    new_broadcast_dimensions = tuple(np.add(1, broadcast_dimensions))
+
+  # TODO(reviewer) This section assumes that the shape of the operand
+  # is broadcast-compatible with the requested shape.  Where should
+  # that be checked, and what should be new rules be, in light of
+  # raggedness?
```

Comment:
We think it should be checked in the abstract eval rule (ie `_broadcast_in_dim_shape_rule` above), which would check the reduced/tracing shape (ie with a variable in it), since shape/type checking usually happens in abstract eval rules. But we're not actually sure if the abstract eval rule generally gets called before the time we reach here.